cmd/image/sync/sync.go

```diff
@@ -0,0 +1,217 @@
+			return err
+		}
+
+		// Unmashalling yaml file
+		var spec []pkg.Spec
+		yamlFile, err := ioutil.ReadFile(opt.SpecYAML)
+		if err != nil {
+			klog.Errorf("ERROR: Read yaml failed : %v", err)
+			return err
+		}
+
+		err = yaml.Unmarshal(yamlFile, &spec)
+		if err != nil {
+			klog.Errorf("Unmarshal: %v", err)
+			return err
+		}
+
+		copyWorker := func(copyJobs <-chan copyWorkload, results chan<- bool, workerId int) {
+			for copyJob := range copyJobs {
+				start := time.Now()
+				klog.Infof("Copying object: %s src bucket: %s dest bucket: %s", copyJob.srcObject, copyJob.srcBucket, copyJob.tgtBucket)
+				err = copyJob.s3Cli.CopyObjectToBucket(copyJob.srcBucket, copyJob.tgtBucket, copyJob.srcObject)
+				if err != nil {
+					klog.Errorf("ERROR: %v, Copy object %s failed", err, copyJob.srcObject)
+					results <- false
+				}
+				duration := time.Since(start)
+				klog.Infof("Copying object: %s from bucket: %s to bucket: %s took %v", copyJob.srcObject, copyJob.srcBucket, copyJob.tgtBucket, duration)
+				results <- true
+			}
+		}
+
+		// Calculating total channels required
+		totalChannels := 0
+		for _, item := range spec {
+			s3Cli, err = client.NewS3Client(bxCli, item.Source.Cos, item.Source.Region)
+			if err != nil {
+				return err
+			}
+
+			_, err = s3Cli.CheckBucketLocationConstraint(item.Source.Bucket, item.Source.Region+"-"+item.Source.StorageClass)
+			if err != nil {
+				klog.Errorf("Location constraint verification failed for src bucket: %s", item.Source.Bucket)
+				return err
+			}
+
+			selectedObjects, err = s3Cli.SelectObjects(item.Source.Bucket, item.Source.Object)
+			if err != nil {
+				klog.Errorf("Select Objects failed: %v", err)
+				return err
+			}
+
+			noOfTargets := len(item.Target)
+			totalChannelsForSrc := noOfTargets * len(selectedObjects)
+			totalChannels = totalChannels + totalChannelsForSrc
+		}
+
+		// Creating workers and channels
+		copyJobs := make(chan copyWorkload, totalChannels)
+		results := make(chan bool, totalChannels)
+		for w := 1; w <= NoOfCopyWorkers; w++ {
+			go copyWorker(copyJobs, results, w)
+		}
+
+		for _, item := range spec {
+			// Creating S3 client
+			s3Cli, err = client.NewS3Client(bxCli, item.Source.Cos, item.Source.Region)
+			if err != nil {
+				return err
+			}
+
+			selectedObjects, err = s3Cli.SelectObjects(item.Source.Bucket, item.Source.Object)
+			if err != nil {
+				klog.Errorf("Select Objects failed: %v", err)
+				return err
+			}
+
+			klog.Infof("Selected Objects from bucket %s: %s", item.Source.Bucket, strings.Join(selectedObjects, ", "))
+			for _, targetItem := range item.Target {
+				s3Cli, err = client.NewS3Client(bxCli, item.Source.Cos, targetItem.Region)
+				if err != nil {
+					return err
+				}
+
+				_, err = s3Cli.CheckBucketLocationConstraint(targetItem.Bucket, targetItem.Region+"-"+targetItem.StorageClass)
+				if err != nil {
+					klog.Errorf("Location constraint verification failed for dest bucket: %s", targetItem.Bucket)
```

Comment:
I see that in line number 149, the totalChannells are calculated as follow
```
			noOfTargets := len(item.Target)
			totalChannelsForSrc := noOfTargets * len(selectedObjects)
			totalChannels = totalChannels + totalChannelsForSrc
```
Also in line number 200, The result channel is read for `totalChannels` times, But here(line number 184) in case of error you are simply continuing to the next item in `item.Target ` and not feeding value to copyJobs

If the copyJobs is not filled for  `totalChannels` times then `result` channel cannot be read for `totalChannels` times, I think this will lead to a deadlock in go routines

Reply:
Correct, it can cause deadlock. I Will return error on failure.